### PR TITLE
base, source: Encapsulate source and error

### PR DESCRIPTION
**Pull request by qinqon**
Encapsulate snippet generation at their own entity source, also add a
factory function to create a reader from it, so the source type is
abstracted. Added also an error that will containe position, message and
snippet.

With this two pieces it's possible to compose an user friendly error that will look like the following

```
invalid equality filter: right hand argument is not a string, pos=28
| routes.running.destination==this.is.wrong                                     
| ............................^`
```

The error is a struct with the fields `position` and `msg`  and optionalia `snippet`, this `snippet` can be filled in with `source.Snippet` function if users want to.

Close https://github.com/nmstate/nmpolicy/issues/46

### CODE REVIEW

File: nmpolicy/internal/source/source.go [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+import (
+	"fmt"
+	"strings"
+)
+
+// Source represents the raw data for a nmpolicy expression
+type Source struct {
+	src    string
+	reader *strings.Reader
```
Comment (AlonaKaplan) [inline nmpolicy/internal/source/source.go:11]: Why the reader is needed? What is the use case of it?

File: nmpolicy/internal/source/error.go [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+// Decorate add the `Snippet` to `Error`, using the Position and error and
+// source it will compose a string with the expression and a pointer.
+func (e *Error) Decorate(src Source) *Error {
+	e.Snippet = src.Snippet(e.Position)
+	return e
+}
+
+// Error mark this struct as a golang error.
+func (e *Error) Error() string {
```
Comment (AlonaKaplan) [inline nmpolicy/internal/source/error.go:23]: Please add unit tests to `Error`.

**Review by AlonaKaplan — COMMENTED** [review]
Please explain in the P message why this PR is needed and what flow in the lexer->parser->resolver will use it.

File: nmpolicy/internal/source/error.go (hunk 2 above)
Comment (qinqon) [reply nmpolicy/internal/source/error.go:23]: Forgot to push the tests file.

**Pull request closed without merging**